Evolution/0016-mutli-producer-single-consumer-channel.md

```diff
@@ -0,0 +1,838 @@
+# MultiProducerSingleConsumerAsyncChannel
+
+* Proposal: [SAA-0016](0016-multi-producer-single-consumer-channel.md)
+* Authors: [Franz Busch](https://github.com/FranzBusch)
+* Status: **Implemented**
+
+## Revision
+- 2023/12/18: Migrate proposal from Swift Evolution to Swift Async Algorithms.
+- 2023/12/19: Add element size dependent strategy
+- 2024/05/19: Rename to multi producer single consumer channel
+- 2024/05/28: Add unbounded strategy
+- 2025/03/24: Adopt `~Copyable` for correct semantics and better performance.
+
+## Introduction
+
+[SE-0314](https://github.com/apple/swift-evolution/blob/main/proposals/0314-async-stream.md)
+introduced new `Async[Throwing]Stream` types which act as root asynchronous
+sequences. These two types allow bridging from synchronous callbacks such as
+delegates to an asynchronous sequence. This proposal adds a new root primitive
+with the goal to model asynchronous multi-producer-single-consumer systems.
+
+## Motivation
+
+After using the `AsyncSequence` protocol, the `Async[Throwing]Stream` types, and
+the `Async[Throwing]Channel` types extensively over the past years, we learned
+that there is a gap in the ecosystem for a type that provides strict
+multi-producer-single-consumer guarantees with external backpressure support.
+Additionally, any stream/channel like type needs to have a clear definition
+about the following behaviors:
+
+1. Backpressure
+2. Multi/single consumer support
+3. Downstream consumer termination
+4. Upstream producer termination
+
+The below sections are providing a detailed explanation of each of those.
+
+### Backpressure
+
+In general, backpressure is the mechanism that prevents a fast producer from
+overwhelming a slow consumer. It helps stability of the overall system by
+regulating the flow of data between different components. Additionally, it
+allows to put an upper bound on resource consumption of a system. In reality,
+backpressure is used in almost all networked applications.
+
+In Swift, asynchronous sequence also have the concept of internal backpressure.
+This modeled by the pull-based implementation where a consumer has to call
+`next` on the `AsyncIterator`. In this model, there is no way for a consumer to
+overwhelm a producer since the producer controls the rate of pulling elements.
+
+However, the internal backpressure of an asynchronous isn't the only
```

Comment:
```suggestion
However, the internal backpressure of an asynchronous sequence isn't the only
```

"asynchronous" what - seems to be missing a word, I think it's meant to be "stream" or "sequence" - and I'd guess sequence from the context